[PyTorch][Tensor] Introduce tensor.dim_order

digantdesai opened the pull request:
Summary:
This is a stride based attribute for a tensor available in Python.

This can help inspect tensors generated using `torch.empty_permuted(.., physical_layout, ...)`, where physical_layout should match the dim_order returned here. `empty_permuted` will be renamed to use dim_order as the param name in the future. And also help Executorch export pipeline with implementing dim_order based tensors.

Differential Revision: D48134476



cc @albanD

kimishpatel commented on test/test_torch.py:
```diff
@@ -7734,6 +7734,20 @@ def test_helper(dim1, dim2, memory_format):
+    def test_dim_order(self):
+        shape = (2, 3, 5, 7)
+
+        self.assertSequenceEqual(torch.empty(shape).dim_order(), (0, 1, 2, 3), seq_type=tuple)
```
can you add some test cases where we create slice of tensor across multiple dims
e.g. t = t[:, 2:5, 7:9]...

digantdesai replied:
Ah taking a slice doesn't change the strides. So Ok to not have this in the tests

kimishpatel commented on test/test_torch.py:
```diff
@@ -7734,6 +7734,20 @@ def test_helper(dim1, dim2, memory_format):
+        shape = (2, 3, 5, 7)
+
+        self.assertSequenceEqual(torch.empty(shape).dim_order(), (0, 1, 2, 3), seq_type=tuple)
+        self.assertSequenceEqual(torch.empty(shape).transpose(0, 1).dim_order(), (1, 0, 2, 3))
```
Dont know, but this one might be worth explaining that transpose permutes those two dims. Thus if you have `t.sizes() = [2, 3, 4, 5]` then `t.tranpose(0, 1).sizes() = [3, 2, 4, 5]` even though underlying memory has not been shuffled. Which is why new dim order is (1, 0, 2, 3).

digantdesai replied:
Just adding a comment saying since this is stride based, expect dim_order to change with this test.

ezyang commented on torch/_tensor.py:
```diff
@@ -1262,6 +1262,54 @@ def to_sparse_coo(self):
+        """
+        if self.layout == torch.strided:
+            return tuple(
+                [
```
you don't need the list here if you have tuple() wrapping around the top

ezyang reviewed (changes requested):
I don't think the implementation of dim_order() is consistent with TensorIterator's behavior.

kimishpatel commented on test/test_torch.py:
```diff
@@ -7734,6 +7734,20 @@ def test_helper(dim1, dim2, memory_format):
+            self.assertSequenceEqual(dim_order, torch.empty_permuted(shape, dim_order).dim_order())
+
+        self.assertRaises(RuntimeError, lambda: torch.empty(shape).to_sparse().dim_order())
+        self.assertRaises(RuntimeError, lambda: torch.empty(shape).to_mkldnn().dim_order())
```
I think chatgpt could hvae reasoned about this way way faster than how long it took me.

ezyang commented on torch/_tensor.py:
```diff
@@ -1318,6 +1318,97 @@ def to_sparse_coo(self):
+        # NB:
+        # - Based on the implementation in TensorIterator.cpp
+        # - Should have similar behavior (esp in ambiguous cases) to,
+        #   torch._prims_common.compute_elementwise_output_logical_to_physical_perm
```
Is there a reason we cannot literally share this code?

digantdesai replied:
Yeah I was thinking about doing the same at first. Then multiple tensors logic in the _prims_common function and given we are implementing a tensor method, I decided to copy and simplify.

I did add tests to make sure these impls are aligned, since that's what we want.

What do you think?

kimishpatel replied:
You can just invoke that function with a single tensor?

digantdesai replied:
Yes, that is exactly what I am doing in the tests.

kimishpatel replied:
Then why can we just not use that as actual impl, instead of copy paste?

digantdesai replied:
Alright, I will just call that function :)

larryliu0820 commented on torch/_tensor.py:
```diff
@@ -1318,6 +1318,97 @@ def to_sparse_coo(self):
+            stride_a = self.stride()[idx_a]
+            stride_b = self.stride()[idx_b]
+
+            if stride_a == 0 or stride_b == 0:
+                return 0
+
+            if stride_a < stride_b:
+                return -1
+
+            if stride_a > stride_b:
+                return 1
+
+            # stride_a == stride_b
+            if shape[idx_a] > shape[idx_b]:
+                return 1
+
+            # Note: this case is hit when all strides are non-zero and equal, and all
+            # dimensions have the same length
+            return 0
```
Nit: these can be combined in a more structured way:
```suggestion
            if stride_a < stride_b:
                return -1
            elif stride_a > stride_b:
                return 1
            # stride_a == stride_b
            elif shape[idx_a] > shape[idx_b]:
                return 1
            else:
            # Note: this case is hit when all strides are non-zero and equal, and all
            # dimensions have the same length
                return 0
```

kimishpatel commented on test/test_torch.py:
```diff
@@ -7735,6 +7735,72 @@ def test_helper(dim1, dim2, memory_format):
+
+                # cases with same strides, and size == 1
+                t = torch.empty(shape)
+                self._check_dim_order_against_prim_impl(torch.empty(shape))
```
I would prefer if we had this test self contained, although assuming prim_impl is well tested this also seems reasonable. However, if we just direclty use prim_impl then these tests becomes moot or just test prim_impl.

digantdesai replied:
Since now we have two impls, the goal of this test is to make sure they align.

kimishpatel reviewed (commented):
For the most part looks good to me. Just check if we can directly use prim impl

ezyang commented on test/test_torch.py:
```diff
@@ -7735,6 +7735,26 @@ def test_helper(dim1, dim2, memory_format):
+        t = torch.empty(shape)
+        self.assertSequenceEqual(t.dim_order(), (0, 1, 2, 3), seq_type=tuple)
+        # transpose doesn't really change the underlying physical memory
+        # so execpting dim_order change to reflect that (like strides)
```
excepting

ezyang commented on torch/_tensor.py:
```diff
@@ -1319,6 +1319,32 @@ def to_sparse_coo(self):
         """
         return self.to_sparse()
 
+    def dim_order(self):
```
You may want to add torch function override support here, look for things like `has_torch_function_unary` in this file

digantdesai committed:
[PyTorch][Tensor] Introduce tensor.dim_order (#106835)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/106835

This is a stride based attribute for a tensor available in Python.

This can help inspect tensors generated using `torch.empty_permuted(.., physical_layout, ...)`, where physical_layout should match the dim_order returned here. `empty_permuted` will be renamed to use dim_order as the param name in the future. And also help Executorch export pipeline with implementing dim_order based tensors.

Importing in the function to avoid circular dependency on torch.Tensor.

Reviewed By: ezyang

Differential Revision: D48134476

fbshipit-source-id: 58513d445931d122af77ac1e04d35eec36cb4c3a
